component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/TOTPAuthenticator.java

```diff
@@ -475,4 +502,184 @@ private boolean isValidTokenLocalUser(int token, String username, Authentication
+			setUserClaimValues(authenticatedUser, username, updatedClaims);
+			String errorMessage = String.format("User account: %s is locked.", authenticatedUser.getUserName());
+			throw new AuthenticationFailedException(errorMessage);
+		} else {
+			updatedClaims
+					.put(TOTPAuthenticatorConstants.TOTP_FAILED_ATTEMPTS_CLAIM, String.valueOf(currentAttempts + 1));
+			setUserClaimValues(authenticatedUser, username, updatedClaims);
+		}
+	}
+
+	private void resetTotpFailedAttempts(AuthenticationContext context) throws AuthenticationFailedException {
+
+		/*
+		Check whether account locking enabled for TOTP to keep backward compatibility.
+		Account locking is not done for federated flows.
+		 */
+		if (!TOTPUtil.isLocalUser(context) || !TOTPUtil.isAccountLockingEnabledForTotp()) {
+			return;
+		}
+		AuthenticatedUser authenticatedUser =
+				(AuthenticatedUser) context.getProperty(TOTPAuthenticatorConstants.AUTHENTICATED_USER);
+		Property[] connectorConfigs = TOTPUtil.getAccountLockConnectorConfigs(authenticatedUser.getTenantDomain());
+
+		// Return if account lock handler is not enabled.
+		for (Property connectorConfig : connectorConfigs) {
+			if ((TOTPAuthenticatorConstants.PROPERTY_ACCOUNT_LOCK_ON_FAILURE.equals(connectorConfig.getName())) &&
+					!Boolean.parseBoolean(connectorConfig.getValue())) {
+				return;
+			}
+		}
+
+		String username = context.getProperty("username").toString();
+		String usernameWithDomain = IdentityUtil.addDomainToName(authenticatedUser.getUserName(),
+				authenticatedUser.getUserStoreDomain());
+		try {
+			UserRealm userRealm = TOTPUtil.getUserRealm(username);
+			UserStoreManager userStoreManager = userRealm.getUserStoreManager();
+
+			// Avoid updating the claims if they are already zero.
+			String[] claimsToCheck = {TOTPAuthenticatorConstants.TOTP_FAILED_ATTEMPTS_CLAIM,
+					TOTPAuthenticatorConstants.FAILED_LOGIN_LOCKOUT_COUNT_CLAIM};
+			Map<String, String> userClaims = userStoreManager.getUserClaimValues(usernameWithDomain, claimsToCheck,
+					UserCoreConstants.DEFAULT_PROFILE);
+			String failedSmsOtpAttempts = userClaims.get(TOTPAuthenticatorConstants.TOTP_FAILED_ATTEMPTS_CLAIM);
+			String failedLoginLockoutCount =
+					userClaims.get(TOTPAuthenticatorConstants.FAILED_LOGIN_LOCKOUT_COUNT_CLAIM);
+
+			if (NumberUtils.isNumber(failedSmsOtpAttempts) && Integer.parseInt(failedSmsOtpAttempts) > 0 ||
+					NumberUtils.isNumber(failedLoginLockoutCount) && Integer.parseInt(failedLoginLockoutCount) > 0) {
+				Map<String, String> updatedClaims = new HashMap<>();
+				updatedClaims.put(TOTPAuthenticatorConstants.TOTP_FAILED_ATTEMPTS_CLAIM, "0");
+				updatedClaims.put(TOTPAuthenticatorConstants.FAILED_LOGIN_LOCKOUT_COUNT_CLAIM, "0");
+				userStoreManager
+						.setUserClaimValues(usernameWithDomain, updatedClaims, UserCoreConstants.DEFAULT_PROFILE);
+			}
+		} catch (UserStoreException e) {
+			log.error("Error while resetting failed TOTP attempts", e);
+			String errorMessage =
```

Comment:
If we are adding one value to a string, we don't usually use `String.format`

Reply:
Fixed in 8307785